### PR TITLE
Added simple server implementation using Boost library

### PR DESCRIPTION
**Pull request by crystalhsieh**
Used Boost tutorials to try and implement a basic server.

### CODE REVIEW

File: server.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+  	static pointer create(boost::asio::io_service& io_service) {
+  	  return pointer(new tcp_connection(io_service));
+  	}
+
+  	tcp::socket& socket() {
+  	  return socket_;
+  	}
+
+    // async_write() serves data to the client
```
Comment (jasonsliu) [inline server.cpp]: This comment belongs next to the function it's referring to!

File: server.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  	    size_t bytes_transferred) {}
+
+    tcp::socket socket_;
+    std::string message_;
+};
+
+class tcp_server {
+  public:
+  	tcp_server (boost::asio::io_service& io_service)
+  	    : acceptor_(io_service, tcp::endpoint(tcp::v4(), 8118)) {
```
Comment (jasonsliu) [inline server.cpp]: Why did we hard code 8118 here if it's already declared as a const variable above?

File: server.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+      start_accept();
+    }
+
+    tcp::acceptor acceptor_;
+};
+
+
+// Main function to run server
+int main() {
```
Comment (jasonsliu) [inline server.cpp:93]: Should we refactor into a different class? Maybe later?

**Review by jasonsliu — CHANGES_REQUESTED** [review]
Looks good as a start!
In general, would not commit binary files.
Also per your suggestion let's go with server.cpp and remove testserver_simpler.cpp

File: server.cpp (hunk 1 above)
Comment (ryanvoong) [reply server.cpp]: i think the comment is for the right function, but the indentation is actually messed up.

**Comment by ryanvoong** [comment]
we should also separate the class declaration and implementation into .h and .cpp files

File: server.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  	tcp_connection(boost::asio::io_service& io_service) : socket_(io_service) {}
+
+  	void handle_write(const boost::system::error_code& error,
+  	    size_t bytes_transferred) {}
+
+    tcp::socket socket_;
+    std::string message_;
+};
+
+class tcp_server {
```
Comment (ryanvoong) [inline server.cpp]: I think `server` is a better name

File: server.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+using boost::asio::ip::tcp;
+
+namespace {
+  const int PORT = 8118;
+}
+
+// shared_ptr and enable_shared_from_this keeps the connection alive
+// as long as there's an operation referring to it.
+class tcp_connection
```
Comment (ryanvoong) [inline server.cpp]: I think `connection` is a better name (for simplicity)

File: server.cpp (hunk 2 above)
Comment (ryanvoong) [inline server.cpp]: Why use IPv4 instead of IPv6?

File: server.cpp (hunk 1 above)
Comment (crystalhsieh) [reply server.cpp]: I agree with jason -- ill put it next to the async_write

File: server.cpp (hunk 2 above)
Comment (crystalhsieh) [reply server.cpp]: you're right - I was playing with it during testing and never changed it.. so I've changed it to PORT

File: server.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
+#include <iostream>
```
Comment (ryanvoong) [inline server.cpp:7]: add a comment to cite sources

File: server.cpp (hunk 3 above)
Comment (crystalhsieh) [reply server.cpp:93]: maybe later when we have greater functionality than just running a server?

File: server.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  	    size_t bytes_transferred) {}
+
+    tcp::socket socket_;
+    std::string message_;
+};
+
+class tcp_server {
+  public:
+  	tcp_server (boost::asio::io_service& io_service)
+  	    : acceptor_(io_service, tcp::endpoint(tcp::v4(), PORT)) {
```
Comment (ryanvoong) [inline server.cpp]: Why use IPv4 instead of IPv6?

**Review by ryanvoong — CHANGES_REQUESTED** [review]
Indentation seems a little inconsistent.
Also we should separate the class declarations and implementations into .h and .cpp files.

File: server.cpp (hunk 4 above)
Comment (crystalhsieh) [reply server.cpp]: done

File: server.cpp (hunk 5 above)
Comment (crystalhsieh) [reply server.cpp]: done

File: server.cpp (hunk 6 above)
Comment (crystalhsieh) [reply server.cpp:7]: done

File: server.cpp (hunk 7 above)
Comment (crystalhsieh) [reply server.cpp]: oops. wasn't really considering pros/cons of either. changed to v6

**Review by ryanvoong — APPROVED** [review]

**Pull request merged**